File: typescript/examples/parallel_vs_sequential.ts
import { Orchestrator, type ToolCall } from "../src/index.js";

async function fetchUser(userId: number): Promise<Record<string, unknown>> {
  await new Promise((resolve) => setTimeout(resolve, 150));
  return { id: userId, name: "Alice", email: "alice@example.com" };
}

async function fetchCatalog(category: string): Promise<Array<Record<string, unknown>>> {
  await new Promise((resolve) => setTimeout(resolve, 150));
  return [
    { sku: "A1", name: "Widget Pro", price: 29.99, category },
    { sku: "A2", name: "Gadget Plus", price: 49.99, category },
  ];
}

async function buildCart(
  user: Record<string, unknown>,
  catalog: Array<Record<string, unknown>>
): Promise<Record<string, unknown>> {
  await new Promise((resolve) => setTimeout(resolve, 100));
  return {
    user: user.name,
    items: [catalog[0]],
    total: catalog[0].price,
  };
}

async function sendReceipt(
  cart: Record<string, unknown>,
  email: string
): Promise<Record<string, unknown>> {
  await new Promise((resolve) => setTimeout(resolve, 50));
  return { sent: true, to: email, total: cart.total };
}

async function runSequential(): Promise<number> {
  const start = performance.now();
  const user = await fetchUser(42);
  const catalog = await fetchCatalog("widgets");
  const cart = await buildCart(user, catalog);
  await sendReceipt(cart, String(user.email));
  return performance.now() - start;
}

async function runWithSynapse(): Promise<number> {
  const orchestrator = new Orchestrator({
    tools: { fetchUser, fetchCatalog, buildCart, sendReceipt },
  });

  const calls: ToolCall[] = [
    { id: "user", name: "fetchUser", inputs: { userId: 42 } },
    { id: "catalog", name: "fetchCatalog", inputs: { category: "widgets" } },
    {
      id: "cart",
      name: "buildCart",
      inputs: { user: "$results.user", catalog: "$results.catalog" },
    },
    {
      id: "receipt",
      name: "sendReceipt",
      inputs: { cart: "$results.cart", email: "$results.user.email" },
    },
  ];

  const start = performance.now();
  const report = await orchestrator.run(calls);
  const elapsed = performance.now() - start;
  console.log(`Speedup estimate: ${report.speedupEstimate.toFixed(2)}x`);
  return elapsed;
}

async function main(): Promise<void> {
  const sequentialMs = await runSequential();
  const synapseMs = await runWithSynapse();
  console.log(`Sequential: ${sequentialMs.toFixed(0)}ms`);
  console.log(`Synapse: ${synapseMs.toFixed(0)}ms`);
  console.log(`Observed speedup: ${(sequentialMs / synapseMs).toFixed(2)}x`);
}

void main();

File: examples/parallel_vs_sequential.py
"""
Example 1 — Parallel vs Sequential Execution
=============================================

Demonstrates how Synapse automatically identifies independent tool calls
and runs them concurrently, while correctly sequencing dependent calls.

Simulated pipeline:
  ┌─────────────┐   ┌─────────────────┐
  │ fetch_user  │   │  fetch_catalog  │  ← no deps, run in parallel
  └──────┬──────┘   └────────┬────────┘
         │                   │
         └─────────┬─────────┘
                   ▼
          ┌─────────────────┐
          │  build_cart     │  ← depends on both above
          └────────┬────────┘
                   ▼
          ┌─────────────────┐
          │  send_receipt   │  ← depends on build_cart
          └─────────────────┘

Without Synapse (sequential): ~600 ms
With Synapse (parallel):       ~300 ms  →  ~2x speedup
"""

import asyncio
import sys
import time

# Allow running the example without installing the package
sys.path.insert(0, "python")

from synapse import Orchestrator, ToolCall

# ── Simulated tool functions (each sleeps to mimic real I/O) ─────────────────

async def fetch_user(user_id: int) -> dict:
    await asyncio.sleep(0.15)  # 150 ms simulated DB call
    return {"id": user_id, "name": "Alice", "email": "alice@example.com"}


async def fetch_catalog(category: str) -> list:
    await asyncio.sleep(0.15)  # 150 ms simulated API call
    return [
        {"sku": "A1", "name": "Widget Pro", "price": 29.99},
        {"sku": "A2", "name": "Gadget Plus", "price": 49.99},
    ]


async def build_cart(user: dict, catalog: list) -> dict:
    await asyncio.sleep(0.10)  # 100 ms build logic
    return {
        "user": user["name"],
        "items": catalog[:1],
        "total": catalog[0]["price"],
    }


async def send_receipt(cart: dict, email: str) -> dict:
    await asyncio.sleep(0.05)  # 50 ms email send
    return {"sent": True, "to": email, "total": cart["total"]}


# ── Sequential baseline ───────────────────────────────────────────────────────

async def run_sequential() -> float:
    print("\n--- Sequential execution (baseline) ---")
    t0 = time.perf_counter()

    user = await fetch_user(42)
    catalog = await fetch_catalog("widgets")
    cart = await build_cart(user, catalog)
    receipt = await send_receipt(cart, user["email"])

    elapsed = (time.perf_counter() - t0) * 1000
    print(f"Receipt sent: {receipt}")
    print(f"Sequential wall-clock: {elapsed:.0f} ms")
    return elapsed


# ── Synapse parallel execution ────────────────────────────────────────────────

async def run_with_synapse() -> float:
    print("\n--- Synapse parallel execution ---")

    orch = Orchestrator(
        tools={
            "fetch_user":    fetch_user,
            "fetch_catalog": fetch_catalog,
            "build_cart":    build_cart,
            "send_receipt":  send_receipt,
        },
        on_call_start=lambda c: print(f"  → starting  [{c.id}] {c.name}"),
        on_call_end=lambda r: print(
            f"  ✓ finished  [{r.call_id}] {r.tool_name} ({r.duration_ms:.0f} ms)"
        ),
    )

    # Print the plan before executing
    plan = orch.plan([
        ToolCall(id="user",    name="fetch_user",    inputs={"user_id": 42}),
        ToolCall(id="catalog", name="fetch_catalog", inputs={"category": "widgets"}),
        ToolCall(id="cart",    name="build_cart",
                 inputs={"user": "$results.user", "catalog": "$results.catalog"}),
        ToolCall(id="receipt", name="send_receipt",
                 inputs={"cart": "$results.cart", "email": "$results.user.email"}),
    ])

    print("\nExecution plan:")
    print(plan)

    t0 = time.perf_counter()
    report = await orch.run([
        ToolCall(id="user",    name="fetch_user",    inputs={"user_id": 42}),
        ToolCall(id="catalog", name="fetch_catalog", inputs={"category": "widgets"}),
        ToolCall(id="cart",    name="build_cart",
                 inputs={"user": "$results.user", "catalog": "$results.catalog"}),
        ToolCall(id="receipt", name="send_receipt",
                 inputs={"cart": "$results.cart", "email": "$results.user.email"}),
    ])
    elapsed = (time.perf_counter() - t0) * 1000

    print(f"\n{report}")
    print(f"Synapse wall-clock: {elapsed:.0f} ms")
    return elapsed


# ── Main ──────────────────────────────────────────────────────────────────────

async def main():
    seq_ms  = await run_sequential()
    par_ms  = await run_with_synapse()

    speedup = seq_ms / par_ms
    print(f"\n{'='*55}")
    print(f"  Sequential : {seq_ms:.0f} ms")
    print(f"  Synapse    : {par_ms:.0f} ms")
    print(f"  Speedup    : {speedup:.2f}x")
    print(f"{'='*55}")


if __name__ == "__main__":
    asyncio.run(main())
